Add a custom log handler and GUI viewer with filters

brisvag opened the pull request:
# References and relevant issues
Followup on https://github.com/napari/napari/pull/6849#issuecomment-2095676630

# Description
The idea is to have our own custom handler for log messages so we can then expose the logs as we prefer from the GUI. This will make it easier for users to copy-paste logs even if they didn't manually open from the console, and also easier for developers to quickly check things.

I never really worked with the logging module, so I have a feeling that this is garbage... but oh well, let's see how bad :P

To test, open napari and do some stuff. Then, from the terminal or console run:
```py
napari._LOG_STREAM.logs_at_level(10)
```

EDIT: Added a very quick-and-dirty GUI for the logger based on the implementation of the `About` widget. For how bad it is, it's actually ok xD You can open it from the menu `Help -> Show log".

cc @czaki

brisvag committed:
simple custom logger implementation

brisvag committed:
ugly logging dialog

Czaki commented on napari/_qt/dialogs/log_dialog.py:
```diff
@@ -0,0 +1,58 @@
+        self.setObjectName('LogDialog')
+        self.setWindowTitle(trans._('napari log'))
+        self.setWindowModality(Qt.WindowModality.ApplicationModal)
+        self.exec_()
```
Exec should not be called in constructor, but in calling place.

Czaki commented on napari/__init__.py:
```diff
@@ -42,6 +44,8 @@
     'viewer': ['Viewer', 'current_viewer'],
 }
 
+_LOG_STREAM = _get_custom_log_stream()
```
I think that custom log handler should be initialized in `napari.qt.qt_event_loop`, same as notification manager.

brisvag replied:
But the logger should work regardless of qt right? Also headless it would be useful.

Czaki replied:
As I understand, it is log handler, not logger. The current code means, that each time one import napari it starts storing all logs in memory, even without a clean way to show them.

brisvag replied:
Yes... but isn't that what we want? Would be good to be able to get these logs programmatically as well. We can also discard old logs if we go above a certain size.

Czaki replied:
My personal feeling is that we should do this, like with notifications (warnigs), but I may be wrong.

brisvag replied:
do what? I'm not sure I follow ^^'

Czaki commented on napari/utils/_logging.py:
```diff
@@ -0,0 +1,49 @@
+    )
+    logger = logging.getLogger('napari')
+    logger.setLevel(logging.DEBUG)
+    logger.addHandler(handler)
```
It will be nice to connect to all loggers and be able to filter messages based on logger (so someone could filter logs only from a given plugin logger.

brisvag replied:
True... I'm not sure how to do that though; I guess we somehow get the root logger? And btw, is there a way to get these logging parameters (level, thread, etc) "raw" instead of formatting to string and then de-formatting?

brisvag replied:
EDIT: I changed to use the root logger and filter for `napari` by default. Probably less finnicky than registering specific plugins.

brisvag committed:
some simple filtering from gui

Czaki committed:
Trigger docs deployment on tagged commit (#7122)

# References and relevant issues

Followup to https://github.com/napari/docs/pull/463

# Description

It looks like the condition from https://github.com/napari/docs/pull/463
was never triggered, as base workflow in repository was triggered only
by push to main branch, not on tag.

Czaki committed:
Fix pinning packages workflow by remove speccial pinning from macOS and add special for Windows (#7119)

# References and relevant issues
fix failing #7045

# Description

Since 2024 July 1 there are wheels for PyQt5-Qt5, not only for macOS,
but also for linux https://pypi.org/project/PyQt5-Qt5/#files

It means that our pinning, that is calculated on linux, no longer works
for windows runners.

This PR removes the special case for macOS, and introduce it for
Windows.

---------

Co-authored-by: Peter Sobolewski <76622105+psobolewskiPhD@users.noreply.github.com>

brisvag committed:
add filtering and better rendering

brisvag committed:
happy mypy

brisvag committed:
fix show

brisvag committed:
Merge branch 'main' of github.com:napari/napari into feature/custom_logger

brisvag committed:
better formatting, use deque to limit amount kept

brisvag committed:
keep log up to date, do not block window

brisvag committed:
happy mypy

brisvag committed:
Merge branch 'main' of github.com:napari/napari into feature/custom_logger

Czaki committed:
Merge branch 'main' into feature/custom_logger

jni committed:
Merge branch 'main' into feature/custom_logger

lucyleeow commented on napari/_qt/_qapp_model/qactions/_help.py:
```diff
@@ -107,4 +112,11 @@ def _show_about(window: Window):
         menus=[{'id': MenuId.MENUBAR_HELP, 'group': MenuGroup.NAVIGATION}],
     ),
+    Action(
+        id='napari.window.help.log',
```
```suggestion
        id='napari.window.help.show_log',
```

very small nitpick, could we match the callback?

Czaki commented on napari/_qt/dialogs/log_dialog.py:
```diff
@@ -0,0 +1,123 @@
+        d.setObjectName('LogDialog')
+        d.setWindowTitle(trans._('napari Log'))
+        d.setModal(False)
+        d.show()
```
I think that this method should produce singleton of logging dialog. So trigger it again should popup window to front instead of creating new one.

brisvag replied:
How would you do that properly in qt to make sure we don't have leaks? Can you suggest a change?

Czaki replied:
class level variable `__instance__` to store weakref to object. If object is alive, and visible (`.isVisible`) then `activateWindow` method could be used to popup to front.

Czaki reviewed (commented):
I need to play more (I need to create code for log genearion - but maybe you have such snippet).


This PR definitelly requires more code coverage.

brisvag committed:
Update napari/_qt/_qapp_model/qactions/_help.py

Co-authored-by: Lucy Liu <jliu176@gmail.com>

brisvag committed:
Merge branch 'main' of github.com:napari/napari into feature/custom_logger

jni commented on napari/__init__.py:
```diff
@@ -42,6 +44,8 @@
     'viewer': ['Viewer', 'current_viewer'],
 }
 
+register_logger('')  # root logger
```
I'm not 100% on logger registration. Should we not use `'napari'` as the logger name here? (I would consider using the root for napari main(), for example, but napari when running from a script.)

jni replied:
> I'm not 100% on logger registration.

ie I'm more like 0%. 😂

brisvag replied:
At first I had `napari` here, and was using this function to also register all plugins... but then what's the cost to simply register anything? what if some info comes from the numpy logger? Or the user's custom code that isn't a plugin?

jni replied:
ooooh!!! This is registering all loggers to *output* to our logger! I thought this was setting the root of all log messages we would emit. 😅 I wonder if we can rename this to `register_log_viewer()` or `register_napari_log_inspector()` or similar? What do you think? Then add a comment:

> make sure that log messages from any library are visible in the napari log inspector

or similar? Did I get that right?

brisvag replied:
Yeah this function basically says "if the logger name starts with `module`, add our handler to it", which results in the log messages being registered and displayed by our logging widget. I renamed it to `register_logger_to_napari_handler` to be as clear as possible. I think the docstring shoudl cover your proposed comment, right? Note that this is not qt or widget specific. We can use this anywhere even without GUI.

brisvag committed:
make it into a dock widget instead

brisvag committed:
better filtering and layout, allow loglevel change

brisvag committed:
add line by line to allow better selection

brisvag committed:
Merge branch 'main' into feature/custom_logger

jni commented on napari/_qt/_qapp_model/qactions/_help.py:
```diff
@@ -11,13 +11,20 @@
+def _show_log(window: Window):
+    window.add_dock_widget(
+        LogWidget(), name='logger', area='bottom', tabify=True
+    )
```
It seems to me like this line could be covered relatively easily?

jni replied:
(ie I would add a qapp test that instantiates the log widget)

jni commented on napari/utils/_logging.py:
```diff
@@ -0,0 +1,98 @@
+        # TODO: actually save log to a file somewhere so it can be retrieved?
+        self.changed()
+
+    def flush(self) -> None:
+        pass
```
What's the dealio here 😂 Maybe add a docstring?

brisvag replied:
Explained in the class docstring (this is required by the superclass).

jni reviewed (changes requested):
@brisvag this is very close! Do you want to write a simple instantiation test (I think all the lines will be covered just by creating a log panel and adding a few log messages as in your test script), and maybe add docstrings to all the functions? ;) Then I think we can merge!

brisvag committed:
Merge branch 'main' of github.com:napari/napari into feature/custom_logger

brisvag committed:
docstrings and small tests

Czaki committed:
Merge branch 'main' into feature/custom_logger

Czaki committed:
Merge branch 'main' into feature/custom_logger

brisvag committed:
clearer name

brisvag committed:
Merge branch 'main' of github.com:napari/napari into feature/custom_logger

brisvag committed:
add workaround for py310

Czaki commented on napari/_qt/widgets/_tests/test_qt_logger.py:
```diff
@@ -0,0 +1,7 @@
+def test_qt_logger(qtbot):
+    widget = LogWidget()
+    qtbot.addWidget(widget)
+    widget.show()
```
```suggestion
    widget.show()
    widget.hide()
```

brisvag replied:
aahhhnnnn

brisvag committed:
Update napari/_qt/widgets/_tests/test_qt_logger.py

Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>

willingc committed:
Merge branch 'main' into feature/custom_logger

willingc commented on napari/utils/_logging.py:
```diff
@@ -0,0 +1,126 @@
+    # in python 3.10 there's no public mapping, and this getLevelName function
+    # is a bit less resilient (e.g: it "works" both ways (name <-> value), and if
+    # invalid values are passed, it just makes up a new level)
+    def get_log_level_value(log_level_name):
```
```suggestion
    def get_log_level_value(log_level_name: str | int ): -> int
```

willingc commented on napari/utils/_logging.py:
```diff
@@ -0,0 +1,126 @@
+        return logging.getLevelName(log_level_name)
+else:
+
+    def get_log_level_value(log_level_name):
```
```suggestion
    def get_log_level_value(log_level_name: str | int ): -> int
```

willingc committed:
Merge branch 'main' into feature/custom_logger

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

brisvag committed:
Apply suggestions from code review

Co-authored-by: Carol Willing <carolcode@willingconsulting.com>

brisvag committed:
Merge branch 'main' of github.com:napari/napari into feature/custom_logger

brisvag committed:
fix annotation

brisvag committed:
Fix typing

brisvag committed:
Merge branch 'main' into feature/custom_logger

brisvag committed:
Merge branch 'main' into feature/custom_logger

brisvag committed:
Merge branch 'main' of github.com:napari/napari into feature/custom_logger

brisvag committed:
Merge branch 'main' into feature/custom_logger

brisvag committed:
Merge branch 'main' into feature/custom_logger

Czaki committed:
connect logger using context manager

Czaki committed:
check if action is problematic

Czaki committed:
skip logger tests

Czaki committed:
add message

Czaki committed:
fix

Czaki committed:
check if not using `NOTSET` solves the problem

Czaki committed:
fix tests

jni committed:
Rename show_log -> show_logs
